### DISCUSSION TITLE
Transparent [or none] Fill for text

### DISCUSSION THREAD

**Discussion by MSKRoss** [ShapeCrawler/ShapeCrawler#625] [Q&A]
I've setup a slide template with a gradient background and so I need to make sure that any text areas I create either have no Fill or have a transparent fill. When I don't specify a Fill color I get a default solid color background. When I try to specify the Alpha portion of a hex color (e.g. "FFFFFF00" with 00 as alpha), the resulting slide is "corrupted" and throws an error in PP when I load the deck. To be complete, I'd need to specify both the Fill and Border as transparent since I don't want either to be visible.

        shapes.AddRectangle(10, 140,1200,300);
        var agendaText = shapes.Last();
        agendaText.Fill.SetColor("FFFFFF00");
        agendaText.TextFrame.Text = "Introductions \nObjective ..."

Since File Type and Alpha properties are read-only, is there some other way to get this effect? Thanks much!

**Comment by wh1t3h47** [comment]
Hello @MSKRoss, I have checked that there is a small difference between transparent and non-transparent generated XML, and thru manual XML manipulation, it should've been possible to make the text box transparent.

```csharp
var innerXml = textBox.SDKOpenXmlElement.InnerXml;

// Realiza a substituição da cor e preenchimento
innerXml = SrgbBgRegex().Replace(innerXml, "${1}000000${2}", 1);
innerXml = GeometryEndRegex().Replace(innerXml, "$1<a:noFill xmlns:a=\"http://schemas.openxmlformats.org/drawingml/2006/main\" />");

textBox.SDKOpenXmlElement.InnerXml = innerXml;
```

The regexes:
```csharp
        [GeneratedRegex("(</a:prstGeom>)")]
        private static partial Regex GeometryEndRegex();
        [GeneratedRegex("(<a:srgbClr val=\").*?(\"[\\s]*/>)")]
        private static partial Regex SrgbBgRegex();
```

By logging the generated XML, I could check that it should have been valid, however, for some reason the library discards the change I made in the XML and returns the same XML it would have without this code.

By the way, I achieved this thinking process thru opening the pptx as zip and examining the XML. I changed one manually in powerpoint to make the textbox transparent and checked the diff between both.

![2024-03-22_18-04](https://github.com/ShapeCrawler/ShapeCrawler/assets/12009178/0f12dee5-ce07-4633-b5b7-cddacd5e1101)

@ashahabov I'd like to ask if there is any way I can make this change reflect perhaps in the final PPTX XML, that is generated in the slides folder.

@MSKRoss If this is urgent for you, you can also try giving a pass into openxml to remove the bg after the pptx is ready, which should work, and then use the openxml generated pptx

**Reply by AdamShakhabov** [reply]
@MSKRoss, @wh1t3h47, the new method [IShapeFill.SetNoFill()](https://github.com/ShapeCrawler/ShapeCrawler/blob/d82e89fc86c87ad9372cdf6e2272235cca9c170f/src/ShapeCrawler/Drawing/IShapeFill.cs#L54). Could you please check whether it works for your needs.

**Reply by lavanyatdev** [reply]
I cannot add text directly. I have to use the textframe.text property of a shape to add text to my slides. now the issue i am facing is that i want no border and no color for the shape. the setnofill() does that but the outline.weight = 0 does not remove the border. also when i do setnofill() the text also becomes invisible. i am unable to change the color of the text separately. please help

```

public static void SetPageNumber(int pagenumber, ISlideShapes shapecollection, IPresentation pres) 
{
    shapecollection.AddRectangle(x: 30, y: 100, 20, 20);
    var addedpgbox = shapecollection.Last();
    addedpgbox.Outline.Weight = 0;
    var pgnopara = addedpgbox.TextFrame.Paragraphs[0];
    pgnopara.Text = pagenumber.ToString();
    addedpgbox.Y = pres.SlideHeight - addedpgbox.Height - 15;
}


```

**Reply by AdamShakhabov** [reply]
@lavanyatdev, thank you for the report.

Please use the recently added method `.SetNoOutline()` to remove outline:

```C#
var pres = new Presentation();
var shapes = pres.Slides[0].Shapes; 
shapes.AddRectangle(x: 30, y: 100, 20, 20);
var addedShape = shapes.Last();

addedShape.Outline.SetNoOutline();
```

See [Prerelease Version](https://github.com/ShapeCrawler/ShapeCrawler?tab=readme-ov-file#prerelease-version) to learn how to use the beta version.

I'm analyzing also the rest issues...

**Reply by AdamShakhabov** [reply]
> when i do setnofill() the text also becomes invisible

I think this effect occurs when the background is white and the default text color is white. Try to change the text color in this way:

```C#
var pres = new Presentation();
var shapes = pres.Slides[0].Shapes; 
shapes.AddRectangle(x: 30, y: 100, 20, 20);
var addedShape = shapes.Last();
addedShape.TextFrame.Text = "Test";
addedShape.Fill.SetNoFill();
const string blackHex = "000000";

foreach (var para in addedShape.TextFrame.Paragraphs)
{
    foreach (var portion in para.Portions)
    {
        portion.Font.Color.Update(blackHex);
    }
}

pres.SaveAs("change-text-color.pptx");

```

**Reply by lavanyatdev** [reply]
yes it worked. i explicitly had to add color to text

**Comment by brunnobb** [comment]
I have a new PR incoming with the AddNoFill() method for Shapes.
Please check if it can be useful for you.
